src/widgets/query-rule-custom-data/query-rule-custom-data.tsx

```diff
@@ -0,0 +1,79 @@
+import React, { render, unmountComponentAtNode } from 'preact-compat';
+import {
+  getContainerNode,
+  prepareTemplateProps,
+  createDocumentationMessageGenerator,
+} from '../../lib/utils';
+import { WidgetFactory } from '../../types';
+import connectQueryRules, {
+  QueryRulesRenderer,
+} from '../../connectors/query-rules/connectQueryRules';
+import CustomData from '../../components/QueryRuleCustomData/QueryRuleCustomData';
+
+export type QueryRuleCustomDataTemplates = {
+  default: string | ((items: object[]) => string);
+};
+
+type QueryRuleCustomDataWidgetParams = {
+  container: string | HTMLElement;
+  templates?: QueryRuleCustomDataTemplates;
+  transformItems?: (items: object[]) => any;
+};
+
+interface QueryRuleCustomDataConnectorWidgetParams
+  extends QueryRuleCustomDataWidgetParams {
+  container: HTMLElement;
+}
+
+type QueryRuleCustomData = WidgetFactory<QueryRuleCustomDataWidgetParams>;
+
+const withUsage = createDocumentationMessageGenerator({
+  name: 'query-rule-custom-data',
+});
+
+const renderer: QueryRulesRenderer<
+  QueryRuleCustomDataConnectorWidgetParams
+> = ({ userData, instantSearchInstance, widgetParams }) => {
+  const { container, templates } = widgetParams;
+
+  const templateProps = prepareTemplateProps({
+    templates,
+    templatesConfig: instantSearchInstance.templatesConfig,
+  });
+
+  render(
+    <CustomData templateProps={templateProps} items={userData} />,
+    container
+  );
+};
+
+const queryRuleCustomData: QueryRuleCustomData = (
+  {
+    container,
+    templates: userTemplates = {},
+    transformItems = items => items,
+  } = {} as QueryRuleCustomDataWidgetParams
+) => {
+  if (!container) {
+    throw new Error(withUsage('The `container` option is required.'));
+  }
+
+  const defaultTemplates = { default: '' };
+  const templates: QueryRuleCustomDataTemplates = {
```

Comment:
Actually I think the `prepareTemplateProps` already do [the merge of the `templates`](https://github.com/algolia/instantsearch.js/blob/60a965df3443d14759657e282172cf3440621cdd/src/lib/utils/prepareTemplateProps.ts#L16-L18).

Reply:
Indeed, I'll simplify this. I wonder if we still need `prepareTemplateProps` actually. Thoughts?

Reply:
We can keep it for now and see what is the impact of change. It does add some information on the object like `useCustomCompileOptions`. Not sure why it's useful though.

Reply:
I removed `prepareTemplateProps` in the last change.

> Not sure why it's useful though.

Exactly my thought too.